yarn-project/noir-contracts/src/contracts/private_token_airdrop_contract/src/main.nr

```diff
@@ -58,33 +61,83 @@ contract PrivateTokenAirdrop {
+        owner: Field,
+    ) -> distinct pub abi::PrivateCircuitPublicInputs {
+        let storage = Storage::init();
+        let mut context = PrivateContext::new(inputs, abi::hash_args([amount, owner]));
+
+        let msg_sender = context.msg_sender();
+        let this_address = context.this_address();
+
+        // If this function is not called by another function in the same contract, the owner must be msg_sender.
+        if msg_sender != this_address {
+            assert(owner == msg_sender);
+        }
+
+        let balance = storage.balances.at(owner);
+        let sum = decrease_balance_max_amount(&mut context, balance, amount, owner);
+
+        // If sum is 0, there are no notes to be burned.
+        assert(sum != 0);
+
+        if sum != amount {
+            // The destroyed notes' sum is not enough. Keep burning.
+            let amount_to_burn = amount - sum;
+            let this = PrivateTokenAirdropContractInterface::at(this_address);
+            let _res = this.burn(&mut context, amount_to_burn, owner);
+        }
+        
+        // Return private circuit public inputs. All private functions need to return this as it is part of the input of the private kernel..
+        context.finish()
+    }
+
     // Transfers `amount` of tokens from `sender` to a `recipient`.
     fn transfer(
         //*********************************/
         // Should eventually be hidden:
         inputs: PrivateContextInputs,
         //*********************************/
         amount: Field, 
-        sender: Field, 
         recipient: Field,
     ) -> distinct pub abi::PrivateCircuitPublicInputs {
         let storage = Storage::init();
-        let mut context = PrivateContext::new(inputs, abi::hash_args([amount, sender, recipient]));
+        let mut context = PrivateContext::new(inputs, abi::hash_args([amount, recipient]));
 
-        // Pick from the set of sender's notes to spend amount.
+        let sender = context.msg_sender();
         let sender_balance = storage.balances.at(sender);
-        spend_notes(&mut context, sender_balance, amount, sender);
 
-        // Creates new note for the recipient.
-        let recipient_balance = storage.balances.at(recipient);
-        send_note(&mut context, recipient_balance, amount, recipient);
+        let sum = decrease_balance_max_amount(&mut context, sender_balance, amount, sender);
+
+        // If sum is 0, there are no notes to be used.
+        assert(sum != 0);
+
+        if sum != amount {
+            // The destroyed notes' sum is not enough for the transfer.
+            // Burn the remaining amount.
+            // We only call burn() when decrease_balance_max_amount() didn't destroy enough notes.
+            let amount_to_burn = amount - sum;
+            let this_address = context.this_address();
+            let this = PrivateTokenAirdropContractInterface::at(this_address);
+            let _res = this.burn(&mut context, amount_to_burn, sender);
```

Comment:
Would it be possible to loop and call burn indefinitely (or, at least until we reach max limits in the kernel circuit)?
The original issue (that this PR links to) seeks to test a recursive function, but calling `burn` once doesn't quite achieve that goal.
It would be great if one of the tests for this `transfer` function could setup a test case where multiple iterations are required to spend enough notes.